File: packages/typescript/openlibx402-langchain/src/payment-tool.ts
/**
 * X402 Payment Tool for LangChain.js
 *
 * Allows LangChain agents to make payments for API access.
 */

import { DynamicStructuredTool } from '@langchain/core/tools';
import { z } from 'zod';
import { Keypair } from '@solana/web3.js';
import { X402AutoClient } from '@openlibx402/client';
import { X402Error } from '@openlibx402/core';

export interface X402PaymentToolOptions {
  walletKeypair: Keypair;
  rpcUrl?: string;
  maxPayment?: string;
  name?: string;
  description?: string;
}

export function createX402PaymentTool(
  options: X402PaymentToolOptions
): DynamicStructuredTool {
  const {
    walletKeypair,
    rpcUrl,
    maxPayment = '1.0',
    name = 'x402_payment',
    description = 'Make an X402 payment to access a paid API endpoint. Input should be a URL to the API endpoint. Returns the API response after successful payment.',
  } = options;

  return new DynamicStructuredTool({
    name,
    description,
    schema: z.object({
      url: z.string().describe('The API endpoint URL to access'),
      method: z
        .string()
        .optional()
        .default('GET')
        .describe('HTTP method (GET, POST, PUT, DELETE)'),
    }),
    func: async ({ url, method = 'GET' }) => {
      const client = new X402AutoClient(walletKeypair, rpcUrl, {
        maxPaymentAmount: maxPayment,
      });

      try {
        const response = await client.fetch(url, { method });
        const data = response.data;
        return typeof data === 'string' ? data : JSON.stringify(data);
      } catch (error) {
        if (error instanceof X402Error) {
          return `Payment error: ${error.code} - ${error.message}`;
        }
        return `Error: ${(error as Error).message}`;
      } finally {
        await client.close();
      }
    },
  });
}

export class X402PaymentTool extends DynamicStructuredTool {
  private walletKeypair: Keypair;
  private rpcUrl?: string;
  private maxPayment: string;

  constructor(options: X402PaymentToolOptions) {
    const {
      walletKeypair,
      rpcUrl,
      maxPayment = '1.0',
      name = 'x402_payment',
      description = 'Make an X402 payment to access a paid API endpoint. Input should be a URL to the API endpoint. Returns the API response after successful payment.',
    } = options;

    super({
      name,
      description,
      schema: z.object({
        url: z.string().describe('The API endpoint URL to access'),
        method: z
          .string()
          .optional()
          .default('GET')
          .describe('HTTP method (GET, POST, PUT, DELETE)'),
      }),
      func: async ({ url, method = 'GET' }) => {
        return this.execute(url, method);
      },
    });

    this.walletKeypair = walletKeypair;
    this.rpcUrl = rpcUrl;
    this.maxPayment = maxPayment;
  }

  private async execute(url: string, method: string = 'GET'): Promise<string> {
    const client = new X402AutoClient(this.walletKeypair, this.rpcUrl, {
      maxPaymentAmount: this.maxPayment,
    });

    try {
      const response = await client.fetch(url, { method });
      return typeof response.data === 'string'
        ? response.data
        : JSON.stringify(response.data);
    } catch (error) {
      if (error instanceof X402Error) {
        return `Payment error: ${error.code} - ${error.message}`;
      }
      return `Error: ${(error as Error).message}`;
    } finally {
      await client.close();
    }
  }
}

File: packages/python/openlibx402-langchain/openlibx402_langchain/payment_tool.py
"""
X402 Payment Tool for LangChain

Allows LangChain agents to make payments for API access.
"""

from typing import Optional
import asyncio

try:
    from langchain.tools import BaseTool
    from pydantic import Field

    LANGCHAIN_AVAILABLE = True
except ImportError:
    LANGCHAIN_AVAILABLE = False
    BaseTool = object  # type: ignore
    Field = lambda **kwargs: None  # type: ignore

try:
    from solders.keypair import Keypair

    SOLDERS_AVAILABLE = True
except ImportError:
    SOLDERS_AVAILABLE = False
    Keypair = None  # type: ignore

from openlibx402_client import X402AutoClient
from openlibx402_core import X402Error


class X402PaymentTool(BaseTool):
    """
    LangChain tool that allows agents to make payments for API access

    Usage:
        wallet_keypair = load_keypair()

        payment_tool = X402PaymentTool(
            wallet_keypair=wallet_keypair,
            name="pay_for_api",
            description="Make payment to access premium API data"
        )

        agent = initialize_agent(
            tools=[payment_tool, ...],
            llm=llm,
        )
    """

    name: str = "x402_payment"
    description: str = (
        "Make an X402 payment to access a paid API endpoint. "
        "Input should be a URL to the API endpoint. "
        "Returns the API response after successful payment."
    )

    wallet_keypair: Keypair = Field(exclude=True)
    rpc_url: Optional[str] = Field(default=None, exclude=True)
    max_payment: Optional[str] = Field(default="1.0", exclude=True)
    allow_local: bool = Field(default=False, exclude=True)

    def __init__(self, **data):
        if not LANGCHAIN_AVAILABLE:
            raise ImportError(
                "LangChain not installed. Install with: pip install langchain"
            )
        if not SOLDERS_AVAILABLE:
            raise ImportError(
                "solders not installed. Install with: pip install solders"
            )
        super().__init__(**data)

    def _run(self, url: str, method: str = "GET", **kwargs) -> str:
        """Synchronous run (calls async version)"""
        return asyncio.run(self._arun(url, method, **kwargs))

    async def _arun(self, url: str, method: str = "GET", **kwargs) -> str:
        """
        Make paid API request

        Args:
            url: API endpoint URL
            method: HTTP method
            **kwargs: Additional request parameters

        Returns:
            API response as string
        """
        client = X402AutoClient(
            wallet_keypair=self.wallet_keypair,
            rpc_url=self.rpc_url,
            max_payment_amount=self.max_payment,
            allow_local=self.allow_local,
        )

        try:
            response = await client.fetch(url, method=method, **kwargs)
            return response.text
        except X402Error as e:
            return f"Payment error: {e.code} - {e.message}"
        except Exception as e:
            return f"Error: {str(e)}"
        finally:
            await client.close()
